### PR TITLE
feat: dimmable lights

### PR DESCRIPTION
**Pull request by znepb**
This PR implements dimmable lights, a redstone lamp-like block, which outputs a light level equivalent to the block's redstone input. This would be pretty cool to have in a cinema or other areas where lighting levels can be perfectly adjusted to set the mood. The intent is that these are controlled by computers or redstone integrators, but they can also be controlled by comparators. 

A sample video can be found here: https://cdn.discordapp.com/attachments/705234278388203601/1113653405433483314/2023-05-31_22-20-21.mp4. The program used in that video can be found here: https://pastebin.com/eFzd6gz9.

This is my first attempt at adding a block with Kotlin, so my code is probably quite questionable. If any problems are discovered, please let me know and I'll try my best to fix them.

### CODE REVIEW

File: src/main/kotlin/io/sc3/goodies/datagen/LanguageProvider.kt [hunk 1]
```diff
@@ -136,6 +136,9 @@ class LanguageProvider(out: FabricDataOutput) : FabricLanguageProvider(out) {
     builder.add(ModBlocks.autumnGrass, "Autumn Grass")
     builder.add(ModBlocks.blueGrass, "Peppy Grass")
 
+    builder.add(ModBlocks.dimmableLight, "Dimmable Light")
+    builder.sub(ModBlocks.dimmableLight, "Emits an amount of light equivalent to the blocks redstone input")
```
Comment (Lemmmy) [inline src/main/kotlin/io/sc3/goodies/datagen/LanguageProvider.kt]:
```suggestion
    builder.sub(ModBlocks.dimmableLight, "Emits an amount of light equivalent to the block's redstone input")
```

File: src/main/kotlin/io/sc3/goodies/misc/DimmableLight.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import net.minecraft.state.StateManager
+import net.minecraft.state.property.IntProperty
+import net.minecraft.state.property.Properties
+import net.minecraft.util.math.BlockPos
+import net.minecraft.world.World
+import java.util.function.ToIntFunction
+
+fun createLightLevelFromPowerState(): ToIntFunction<BlockState> {
+  return ToIntFunction { state: BlockState -> state.get(Properties.POWER) ?: 0 }
+}
```
Comment (Lemmmy) [inline src/main/kotlin/io/sc3/goodies/misc/DimmableLight.kt]: Move this to be inside the companion object in `DimmableLight`, and use it like `DimmableLight.createLightLevel...` in `Registration`

File: src/main/kotlin/io/sc3/goodies/misc/DimmableLight.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    sourcePos: BlockPos?,
+    notify: Boolean
+  ) {
+    if (!world!!.isClient) {
+      val receivedPower = world.getReceivedRedstonePower(pos)
+      if (state!!.get(power) != receivedPower) {
+        world.setBlockState(pos, state.with(power, receivedPower), 2)
+      }
+    }
+  }
```
Comment (Lemmmy) [inline src/main/kotlin/io/sc3/goodies/misc/DimmableLight.kt]: None of these parameters need to be nullable (`?`, see [RedstoneIntegratorBlock](https://github.com/SwitchCraftCC/Plethora-Fabric/blob/19d2192551696fff717edc62eea314ad0d236a59/src/main/java/io/sc3/plethora/gameplay/redstone/RedstoneIntegratorBlock.kt#L31)), and you will be able to remove all the not-null assertions (`!!`) when you fix that.

General Kotlin tip: always treat not-null assertions with suspicion, as it is the language's way of explicitly telling you that a null-pointer exception can occur here. You should never need them.

**Review by Lemmmy — CHANGES_REQUESTED** [review]
Thanks for the PR! The code generally looks good. I've made a few formatting changes, and requested a few small further changes here. 

I think that the textures look a little flat and don't fit in too well with the rest of the vanilla textures. Are they placeholder textures? I would be happy to take a look at doing those if you'd like.

File: src/main/kotlin/io/sc3/goodies/misc/DimmableLight.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    world: World?,
+    pos: BlockPos?,
+    sourceBlock: Block?,
+    sourcePos: BlockPos?,
+    notify: Boolean
+  ) {
+    if (!world!!.isClient) {
+      val receivedPower = world.getReceivedRedstonePower(pos)
+      if (state!!.get(power) != receivedPower) {
+        world.setBlockState(pos, state.with(power, receivedPower), 2)
```
Comment (Lemmmy) [inline src/main/kotlin/io/sc3/goodies/misc/DimmableLight.kt]: Where does this magic number `2` come from? The bitmask for block state changes can be found from line 106 of `Block.java`. A value of 2 is equivalent to NOTIFY_LISTENERS:

```suggestion
        world.setBlockState(pos, state.with(power, receivedPower), Block.NOTIFY_LISTENERS)
```

**Review by Lemmmy — CHANGES_REQUESTED** [review]
Missed a comment before

**Review by Lemmmy — APPROVED** [review]

**Comment by Lemmmy** [comment]
https://github.com/SwitchCraftCC/sc-goodies/assets/858456/e19b1a6e-8c26-49ed-a3d3-a46f80f3a247

Happy with this if you are!

**Comment by znepb** [comment]
Looks great!

**Pull request merged**